File: app/src/main/java/com/example/liftoff/extractor/email/ImapGmailExtractor.java
package com.example.liftoff.extractor.email;

import com.example.liftoff.storage.Storage;
import jakarta.mail.*;
import jakarta.mail.internet.MimeBodyPart;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Stream;

/**
 * Implements necessary logic to connect and download attachments from the Gmail address
 */
public class ImapGmailExtractor extends EmailExtractor {
    private Store store;

    public ImapGmailExtractor(String emailAddress, String userToken, Storage storage) {
        super(emailAddress, userToken, storage);
    }

    private boolean containsMultipartSafe(final Message message) {
        try {
            return message.getContentType().contains("multipart");
        } catch (MessagingException e) {
            System.err.println("Failed to access mail content: " + e.getMessage());
            return false;
        }
    }

    private Optional<Email> messageToEmailSafe(final Message message) {
        try {
            final var attachments = new ArrayList<MimeBodyPart>();
            final var multipartContent = (Multipart) message.getContent();
            for (int i = 0; i < multipartContent.getCount(); i++) {
                final var mimeBodyPart = (MimeBodyPart) multipartContent.getBodyPart(i);
                if (!MimeBodyPart.ATTACHMENT.equalsIgnoreCase(mimeBodyPart.getDisposition())) {
                    continue;
                }
                attachments.add(mimeBodyPart);
            }
            if (attachments.isEmpty()) {
                return Optional.empty();
            }
            return Optional.of(new Email(Optional.ofNullable(message.getSubject()), message.getReceivedDate(), attachments));
        } catch (MessagingException | IOException e) {
            System.err.println("Failed to access mail content: " + e.getMessage());
            return Optional.empty();
        }

    }

    private void extractAttachmentSafe(final DirectoryAttachmentPair pair) {
        try (final var attachmentStream = pair.attachment().getInputStream()) {
            this.storage.storeFile(
                    attachmentStream,
                    pair.attachment().getFileName(),
                    Optional.of(pair.attachmentDirectoryName())
            );
        } catch (IOException e) {
            System.err.println("Failed to store file to " + pair.attachmentDirectoryName() + " " + e.getClass() + ": " + e.getMessage());
        } catch (MessagingException e) {
            System.err.println("Failed to access mail content: " + e.getMessage());
        }

    }

    private String buildAttachmentDirectoryName(final Date receivedDate, final Optional<String> subject) {
        final var receivedDateFormatted = new SimpleDateFormat("dd.MM.yyyy_HH:mm:ss").format(receivedDate);
        return receivedDateFormatted + "_" + subject.orElse("None");
    }

    private Stream<DirectoryAttachmentPair> flattenEmail(final Email email) {
        return email.attachments().stream().map(attachment -> {
            final var attachmentDirectoryName = buildAttachmentDirectoryName(email.received(), email.subject());
            return new DirectoryAttachmentPair(attachmentDirectoryName, attachment);
        });
    }

    @Override
    public void extract() {
        if (!this.isAuthenticated) {
            System.err.println("Cannot extract emails without authentication. Aborting");
            return;
        }
        try {
            final var folder = this.store.getFolder("INBOX");
            folder.open(Folder.READ_ONLY);
            Arrays
                    .stream(folder.getMessages())
                    .limit(20) // For testing only
                    .parallel()
                    .filter(this::containsMultipartSafe)
                    .map(this::messageToEmailSafe)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .flatMap(this::flattenEmail)
                    .forEach(this::extractAttachmentSafe);
        } catch (MessagingException e) {
            System.err.println("Failed to access mail provider: " + e.getMessage());
        }
    }

    @Override
    public void authenticate() throws MessagingException {
        Properties properties = new Properties();
        properties.put("mail.imaps.host", "imap.gmail.com");
        properties.put("mail.imaps.ssl.trust", "imap.gmail.com");
        properties.put("mail.imaps.port", "993");
        properties.put("mail.imaps.starttls.enable", "true");
        properties.put("mail.imaps.connectiontimeout", "10000");
        properties.put("mail.imaps.timeout", "10000");
        final var session = Session.getInstance(properties);
        this.store = session.getStore("imaps");
        this.store.connect("imap.gmail.com", this.emailAddress, this.userToken);
        this.isAuthenticated = true;
    }

    @Override
    public void close() throws MessagingException {
        this.store.close();
    }
}

File: src/main/scala/com/example/liftoff/extractor/email/ImapGmailExtractor.scala
package com.example.liftoff.extractor.email

import com.example.liftoff.extractor.{ExtractorAuthError, ExtractorConnectionCloseError, ExtractorConnectionError, ExtractorError}
import com.example.liftoff.storage.Storage
import jakarta.mail._
import jakarta.mail.internet.MimeBodyPart

import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Properties
import scala.collection.parallel.CollectionConverters.ImmutableIterableIsParallelizable
import scala.util.{Failure, Success, Using}

class ImapGmailExtractor(emailAddress: String, userToken: String, storage: Storage) extends EmailExtractor(emailAddress: String, userToken: String, storage: Storage) {
  private var store: Store = _

  private def containsMultipart(message: Message): Boolean = {
    try {
      message.getContentType.contains("multipart")
    } catch {
      case e: MessagingException =>
        Console.err.println(s"Failed to access multipart content: ${e.getMessage}")
        false
    }
  }

  private def toEmail(message: Message): Option[Email] = {
    try {
      val multipartContent: Multipart = message.getContent.asInstanceOf[Multipart]
      val attachments = (for (i <- 0 until multipartContent.getCount)
        yield multipartContent.getBodyPart(i).asInstanceOf[MimeBodyPart])
        .filter(mimeBodyPart => "attachment".equalsIgnoreCase(mimeBodyPart.getDisposition))
        .toList
      if (attachments.isEmpty) return None
      Some(Email(Option(message.getSubject), message.getReceivedDate, attachments))
    } catch {
      case e: MessagingException =>
        Console.err.println(s"Failed to access multipart content: ${e.getMessage}")
        None
      case e: IOException =>
        Console.err.println(s"Failed to access multipart content: ${e.getMessage}")
        None
    }
  }

  private def extractAttachment(directory_name: String, attachment: MimeBodyPart): Unit = {
    Using(attachment.getInputStream) { attachmentStream =>
      this.storage.storeFile(
        attachmentStream,
        attachment.getFileName,
        Some(directory_name))
    } match {
      case Failure(e) => Console.err.println(s"Failed to store file to $directory_name ${e.getClass}: ${e.getMessage}")
      case Success(_) =>
    }
  }

  private def flattenEmail(email: Email) = {
    email.attachments.map(attachment => {
      val dateReceived = new SimpleDateFormat("dd.MM.yyyy_HH:mm:ss").format(email.received)
      val attachmentDirectoryName = s"${dateReceived}_${email.subject.getOrElse("None")}"
      (attachmentDirectoryName, attachment)
    })
  }

  override def extract(): Option[ExtractorError] = {
    if (!this.isAuthenticated) {
      return Some(ExtractorAuthError)
    }
    val folder = this.store.getFolder("INBOX")
    folder.open(Folder.READ_ONLY)
    folder
      .getMessages
      .toList
      .par
      .take(20)
      .filter(containsMultipart)
      .flatMap(toEmail)
      .flatMap(flattenEmail)
      .foreach { case (directory_name, attachment) => extractAttachment(directory_name, attachment) }
    None
  }

  override def authenticate(): Option[ExtractorError] = {
    val properties = new Properties()
    properties.put("mail.imaps.host", "imap.gmail.com")
    properties.put("mail.imaps.ssl.trust", "imap.gmail.com")
    properties.put("mail.imaps.port", "993")
    properties.put("mail.imaps.starttls.enable", "true")
    properties.put("mail.imaps.connectiontimeout", "10000")
    properties.put("mail.imaps.timeout", "10000")
    val session = Session.getInstance(properties)
    try {
      this.store = session.getStore("imaps")
      this.store.connect("imap.gmail.com", this.emailAddress, this.userToken)
    } catch {
      case e: MessagingException => return Some(ExtractorConnectionError(e.getMessage))
    }
    this.isAuthenticated = true
    None
  }

  override def close(): Option[ExtractorError] = {
    try {
      this.store.close()
    } catch {
      case e: MessagingException => return Some(ExtractorConnectionCloseError(e.getMessage))
    }
    None
  }

}
